File: src/lineardisc.c
#include <R.h>
#include <R_ext/Utils.h>
#include "chunkloop.h"

/* 
   lineardisc.c

   Disc of radius r in linear network

   $Revision: 1.10 $  $Date: 2013/05/27 02:09:10 $

*/

#define DPATH(I,J) dpath[(J) + Nv * (I)]

#include "yesno.h"

#undef DEBUG

void 
lineardisc(f, seg, /* centre of disc (local coords) */
	   r,      /* radius of disc */
	   nv, xv, yv,   /* network vertices */
	   ns, from, to,  /* segments */
	   dpath,  /* shortest path distances between vertices */
	   lengths, /* segment lengths */
	   allinside, boundary, dxv, nendpoints)
     int *nv, *ns;
     int *from, *to; /* integer vectors (mappings) */
     double *f, *r; 
     int *seg;
     double *xv, *yv; /* vectors of coordinates of vertices */
     double *dpath; /* matrix of shortest path distances between vertices */
     double *lengths; /* vector of segment lengths */
     /* OUTPUTS */
     int *allinside, *boundary; /* vectors of status for each segment */
     double *dxv; /* vector of distances for each vertex */
     int *nendpoints;
{
  int Nv, Ns;
  double f0, rad;
  int seg0;

  int i, A, B, fromi, toi, allin, bdry, reachable, nends, maxchunk;
  double length0, dxA, dxB, dxAvi, dxBvi, residue;
  double *resid; 
  int *covered;

  Nv = *nv;
  Ns = *ns;

  f0 = *f;
  seg0 = *seg;
  rad = *r;

  /* endpoints of segment containing centre */
  A = from[seg0];
  B = to[seg0];

  /* distances from x to  A and B */
  length0 = lengths[seg0];
  dxA = f0 * length0;
  dxB = (1-f0) * length0;

  /* visit vertices */
  covered = (int *) R_alloc((size_t) Nv, sizeof(int));
  resid = (double *) R_alloc((size_t) Nv, sizeof(double));

  OUTERCHUNKLOOP(i, Nv, maxchunk, 16384) {
    R_CheckUserInterrupt();
    INNERCHUNKLOOP(i, Nv, maxchunk, 16384) {
      /* distance going through A */
      dxAvi = dxA + DPATH(A,i);
      /* distance going through B */
      dxBvi = dxB + DPATH(B,i);
      /* shortest path distance to this vertex */
      dxv[i] = (dxAvi < dxBvi) ? dxAvi : dxBvi;
      /* distance left to 'spend' from this vertex */
      residue = rad - dxv[i];
      resid[i] = (residue > 0)? residue : 0;
      /* determine whether vertex i is inside the disc of radius r */
      covered[i] = (residue >= 0);
    }
  }
  /* 
     Now visit line segments. 
  */
  nends = 0;

  OUTERCHUNKLOOP(i, Ns, maxchunk, 16384) {
    R_CheckUserInterrupt();
    INNERCHUNKLOOP(i, Ns, maxchunk, 16384) {
      /* 
	 Determine which line segments are completely inside the disc,
	 and which cross the boundary.
      */
      if(i == seg0) {
	/* initial segment: disc starts from centre (x, y) */
	allin = covered[A] && covered[B];
	bdry  = !allin;
	if(bdry) {
	  if(!covered[A]) nends++;
	  if(!covered[B]) nends++;
	}
      } else {
	/* another segment: disc extends in from either endpoint */
	fromi = from[i];
	toi   = to[i];
	reachable = (covered[fromi] || covered[toi]);
	if(reachable) {
	  allin = covered[fromi] && covered[toi] && 
                     (resid[fromi] + resid[toi] >= lengths[i]);
	  bdry = !allin;
	} else allin = bdry = NO;
	if(bdry) {
	  if(covered[fromi]) nends++;
	  if(covered[toi]) nends++;
	}
      }
      allinside[i] = allin;
      boundary[i] = bdry;
    }
  }
  *nendpoints = nends;
}

/* ------------------------------------------------- */
/*   count endpoints of several discs in a network   */
/* ------------------------------------------------- */

void 
Ccountends(np, f, seg, /* centres of discs (local coords) */
	  r,                /* radii of discs */
	  nv, xv, yv,   /* network vertices */
	  ns, from, to,  /* network segments */
	  dpath,  /* shortest path distances between vertices */
	  lengths, /* segment lengths */
	  toler, /* tolerance */
	  nendpoints /* output counts of endpoints */
	  )
     int *np, *nv, *ns;
     int *from, *to; /* integer vectors (mappings) */
     double *f, *r; 
     int *seg;
     double *xv, *yv; /* vectors of coordinates of vertices */
     double *dpath; /* matrix of shortest path distances between vertices */
     double *lengths; /* vector of segment lengths */
     double *toler; /* tolerance for merging endpoints and vertices */
     /* OUTPUT */
     int *nendpoints;
{
  int Np, Nv, Ns;
  double f0, rad;
  int seg0;

  int i, m, A, B, fromi, toi, reachable, nends, maxchunk, covfrom, covto, allin;
  double length0, dxA, dxB, dxAvi, dxBvi, dxvi, residue, resfrom, resto, tol;
  double *resid; 
  int *covered, *terminal;

  Np = *np;
  Nv = *nv;
  Ns = *ns;
  tol = *toler;

  covered = (int *) R_alloc((size_t) Nv, sizeof(int));
  terminal = (int *) R_alloc((size_t) Nv, sizeof(int));
  resid = (double *) R_alloc((size_t) Nv, sizeof(double));

  /* loop over centre points */
  OUTERCHUNKLOOP(m, Np, maxchunk, 256) {
    R_CheckUserInterrupt();
    INNERCHUNKLOOP(m, Np, maxchunk, 256) {

      f0 = f[m];
      seg0 = seg[m];
      rad = r[m];

#ifdef DEBUG
      Rprintf("\nCentre point %d lies in segment %d\n", m, seg0);
#endif

      /* endpoints of segment containing centre */
      A = from[seg0];
      B = to[seg0];

      /* distances from centre to A and B */
      length0 = lengths[seg0];
      dxA = f0 * length0;
      dxB = (1-f0) * length0;

      nends = 0;

      /* visit vertices */
      for(i = 0; i < Nv; i++) {
	/* distance going through A */
	dxAvi = dxA + DPATH(A,i);
	/* distance going through B */
	dxBvi = dxB + DPATH(B,i);
	/* shortest path distance to this vertex */
	dxvi = (dxAvi < dxBvi) ? dxAvi : dxBvi;
	/* distance left to 'spend' from this vertex */
	residue = rad - dxvi;
	if(residue > tol) {
	  resid[i] = residue;
	  covered[i] = YES;
	  terminal[i] = NO;
	} else if(residue < -tol) {
	  resid[i] = 0;
	  covered[i] = terminal[i] = NO;
	} else {
	  /* vertex is within 'tol' of an endpoint 
	   - deem it to be one 
	  */
	  resid[i] = 0;
	  covered[i] = terminal[i] = YES;
	  /* vertex is an endpoint of disc */
	  ++nends;  
	}
      }

#ifdef DEBUG
      Rprintf("%d terminal endpoints\n", nends);
#endif

      /* 
	 Now visit line segments 
	 to count any endpoints that are interior to the segments.
      */

      for(i = 0; i < Ns; i++) {
	/* 
	   Determine which line segments are completely inside the disc,
	   and which cross the boundary.
	*/
	if(i == seg0) {
	  /* initial segment: disc starts from (x0, y0) */
	  if(!covered[A]) nends++;
	  if(!covered[B]) nends++;
#ifdef DEBUG
	  if(!covered[A]) Rprintf("A not covered\n");
	  if(!covered[B]) Rprintf("B not covered\n");
#endif
	} else {
	  /* another segment: disc extends in from either endpoint */
	  fromi = from[i];
	  toi   = to[i];
	  covfrom = covered[fromi];
	  covto   = covered[toi];
	  resfrom = resid[fromi];
	  resto   = resid[toi];
	  reachable = covfrom || covto;
#ifdef DEBUG
	  residue = resfrom + resto - lengths[i];
	  Rprintf("%d: %s %s: %lf + %lf - %lf = %lf sign %s\n", 
		  i,
		  (terminal[fromi]) ? "T" : ((covfrom) ? "Y" : "N"),
		  (terminal[toi]) ? "T" : ((covto) ? "Y" : "N"),
		  resfrom, resto, lengths[i], residue,
		  (residue < 0) ? "-" : ((residue > 0) ? "+" : "0"));
#endif
	  if(reachable) {
	    residue = resfrom + resto - lengths[i];
	    allin = covfrom && covto && (residue >= 0);
#ifdef DEBUG
	    if(allin) {
	      Rprintf("Covered\n"); 
	    } else if((terminal[fromi] || terminal[toi]) &&
		      (residue >= - tol * lengths[i])) {
		Rprintf("Deemed to be covered\n"); 
	    } else Rprintf("Reachable\n");
#endif
	    allin = allin || 
	      ((terminal[fromi] || terminal[toi]) &&
	       (residue >= - tol));
	    if(!allin) {
	      /* segment is not entirely covered by disc
		 - infer endpoint(s) in interior of segment */
	      if(covfrom && !terminal[fromi]) nends++;
	      if(covto && !terminal[toi]) nends++;
#ifdef DEBUG
	      if(covfrom && !terminal[fromi]) Rprintf("fromi => end\n");
	      if(covto && !terminal[toi]) Rprintf("toi => end\n");
#endif
	    }
	  }
	}
      }
      nendpoints[m] = nends;
    }
  }
}

File: R/lineardisc.R
#
#
#   disc.R
#
#   $Revision: 1.28 $ $Date: 2018/03/23 01:26:43 $
#
#   Compute the disc of radius r in a linear network
#
#   
lineardisc <- function(L, x=locator(1), r, plotit=TRUE,
                       cols=c("blue", "red", "green")) {
  # L is the linear network (object of class "linnet")
  # x is the centre point of the disc
  # r is the radius of the disc
  #
  stopifnot(inherits(L, "linnet"))
  check.1.real(r)
  if(L$sparse) {
    message("Converting linear network to non-sparse representation..")
    L <- as.linnet(L, sparse=FALSE)
  }
  lines <- L$lines
  vertices <- L$vertices
  lengths <- lengths.psp(lines)
  win <- L$window
  #
  # project x to nearest segment
  if(missing(x))
    x <- clickppp(1, win, add=TRUE)
  else
    x <- as.ppp(x, win)
  pro <- project2segment(x, lines)
  # which segment?
  startsegment <- pro$mapXY
  # parametric position of x along this segment
  startfraction <- pro$tp
  # vertices at each end of this segment
  A <- L$from[startsegment]
  B <- L$to[startsegment]
  # distances from x to  A and B
  dxA <- startfraction * lengths[startsegment]
  dxB <- (1-startfraction) * lengths[startsegment]
  # is r large enough to reach both A and B?
  startfilled <- (max(dxA, dxB) <= r)
  # compute vector of shortest path distances from x to each vertex j,
  # going through A:
  dxAv <- dxA + L$dpath[A,]
  # going through B:
  dxBv <- dxB + L$dpath[B,]
  # going either through A or through B:
  dxv <- pmin.int(dxAv, dxBv)
  # Thus dxv[j] is the shortest path distance from x to vertex j.
  #
  # Determine which vertices are inside the disc of radius r
  covered <- (dxv <= r)
  # Thus covered[j] is TRUE if the j-th vertex is inside the disc.
  #
  # Determine which line segments are completely inside the disc
  #
  from <- L$from
  to   <- L$to
  # ( a line segment is inside the disc if the shortest distance
  #   from x to one of its endpoints, plus the length of the segment,
  #   is less than r ....
  allinside <- (dxv[from] + lengths <= r) | (dxv[to] + lengths <= r)
  #   ... or alternatively, if the sum of the
  #   two residual distances exceeds the length of the segment )
  residfrom <- pmax.int(0, r - dxv[from])
  residto   <- pmax.int(0, r - dxv[to])
  allinside <- allinside | (residfrom + residto >= lengths)
  # start segment is special
  allinside[startsegment] <- startfilled
  # Thus allinside[k] is TRUE if the k-th segment is inside the disc
  
  # Collect all these segments
  disclines <- lines[allinside]
  #
  # Determine which line segments cross the boundary of the disc
  boundary <- (covered[from] | covered[to]) & !allinside
  # For each of these, calculate the remaining distance at each end
  resid.from <- ifelseXB(boundary, pmax.int(r - dxv[from], 0), 0)
  resid.to   <- ifelseXB(boundary, pmax.int(r - dxv[to],   0), 0)
  # Where the remaining distance is nonzero, create segment and endpoint
  okfrom <- (resid.from > 0)
  okfrom[startsegment] <- FALSE
  if(any(okfrom)) {
    v0 <- vertices[from[okfrom]]
    v1 <- vertices[to[okfrom]]
    tp <- (resid.from/lengths)[okfrom]
    vfrom <- ppp((1-tp)*v0$x + tp*v1$x,
                 (1-tp)*v0$y + tp*v1$y,
                 window=win)
    extralinesfrom <- as.psp(from=v0, to=vfrom)
  } else vfrom <- extralinesfrom <- NULL
  #
  okto <- (resid.to > 0)
  okto[startsegment] <- FALSE
  if(any(okto)) {
    v0 <- vertices[to[okto]]
    v1 <- vertices[from[okto]]
    tp <- (resid.to/lengths)[okto]
    vto <- ppp((1-tp)*v0$x + tp*v1$x,
               (1-tp)*v0$y + tp*v1$y,
               window=win)
    extralinesto <- as.psp(from=v0, to=vto)
  } else vto <- extralinesto <- NULL
  #
  # deal with special case where start segment is not fully covered
  if(!startfilled) {
    vA <- vertices[A]
    vB <- vertices[B]
    rfrac <- r/lengths[startsegment]
    tleft <- pmax.int(startfraction-rfrac, 0)
    tright <- pmin.int(startfraction+rfrac, 1)
    vleft <- ppp((1-tleft) * vA$x + tleft * vB$x,
                 (1-tleft) * vA$y + tleft * vB$y,
                 window=win)
    vright <- ppp((1-tright) * vA$x + tright * vB$x,
                  (1-tright) * vA$y + tright * vB$y,
                  window=win)
    startline <- as.psp(from=vleft, to=vright)
    startends <- superimpose(if(!covered[A]) vleft else NULL,
                             if(!covered[B]) vright else NULL)
  } else startline <- startends <- NULL
  #
  # combine all lines
  disclines <- superimpose(disclines,
                           extralinesfrom, extralinesto, startline,
                           W=win, check=FALSE)
  # combine all disc endpoints
  discends <- superimpose(vfrom, vto, vertices[dxv == r], startends,
                          W=win, check=FALSE)
  #
  if(plotit) {
    if(dev.cur() == 1) {
      # null device - initialise a plot
      plot(L, main="")
    }
    points(x, col=cols[1L], pch=16)
    plot(disclines, add=TRUE, col=cols[2L], lwd=2)
    plot(discends, add=TRUE, col=cols[3L], pch=16)
  }
  return(list(lines=disclines, endpoints=discends))
}

countends <- function(L, x=locator(1), r, toler=NULL) {
  # L is the linear network (object of class "linnet")
  # x is the centre point of the disc
  # r is the radius of the disc
  #
  stopifnot(inherits(L, "linnet"))
  sparse <- L$sparse %orifnull% is.null(L$dpath)
  if(sparse)
    stop(paste("countends() does not support linear networks",
               "that are stored in sparse matrix format.",
               "Please convert the data using as.linnet(sparse=FALSE)"),
         call.=FALSE)
  # get x
  if(missing(x))
    x <- clickppp(1, Window(L), add=TRUE)
  if(!inherits(x, "lpp"))
    x <- as.lpp(x, L=L)
  np <- npoints(x)
  
  if(length(r) != np)
    stop("Length of vector r does not match number of points in x")
  #
  if(!is.connected(L)) {
    #' disconnected network - split into components
    result <- numeric(np)
    lab <- connected(L, what="labels")
    subsets <- split(seq_len(nvertices(L)), factor(lab))
    for(subi in subsets) {
      xi <- thinNetwork(x, retainvertices=subi)
      witch <- which(attr(xi, "retainpoints"))
      ok <- is.finite(r[witch])
      witchok <- witch[ok]
      result[witchok] <-
        countends(domain(xi), xi[ok], r[witchok], toler=toler)      
    }
    return(result)
  }
  lines <- L$lines
  vertices <- L$vertices
  lengths <- lengths.psp(lines)
  dpath <- L$dpath
  nv <- vertices$n
  ns <- lines$n
  #
  if(!spatstat.options("Ccountends")) {
    #' interpreted code
    result <- integer(np)
    for(i in seq_len(np)) 
      result[i] <- npoints(lineardisc(L, x[i], r[i], plotit=FALSE)$endpoints)
    return(result)
  }
  # extract coordinates
  coo <- coords(x)
  #' which segment
  startsegment <- coo$seg 
  # parametric position of x along this segment
  startfraction <- coo$tp
  # convert indices to C 
  seg0 <- startsegment - 1L
  from0 <- L$from - 1L
  to0   <- L$to - 1L
  # determine numerical tolerance
  if(is.null(toler)) {
    toler <- default.linnet.tolerance(L)
  } else {
    check.1.real(toler)
    stopifnot(toler > 0)
  }
  zz <- .C("Ccountends",
           np = as.integer(np),
           f = as.double(startfraction),
           seg = as.integer(seg0),
           r = as.double(r), 
           nv = as.integer(nv), 
           xv = as.double(vertices$x),
           yv = as.double(vertices$y),  
           ns = as.integer(ns),
           from = as.integer(from0),
           to = as.integer(to0), 
           dpath = as.double(dpath),
           lengths = as.double(lengths),
           toler=as.double(toler),
           nendpoints = as.integer(integer(np)),
           PACKAGE = "spatstat")
  zz$nendpoints
}

default.linnet.tolerance <- function(L) {
  # L could be a linnet or psp
  if(!is.null(toler <- L$toler)) return(toler)
  len2 <- lengths.psp(as.psp(L), squared=TRUE)
  len2pos <- len2[len2 > 0]
  toler <- if(length(len2pos) == 0) 0 else (0.001 * sqrt(min(len2pos)))
  toler <- makeLinnetTolerance(toler)
  return(toler)
}

makeLinnetTolerance <- function(toler) {
  max(sqrt(.Machine$double.xmin),
      toler[is.finite(toler)], na.rm=TRUE)
}




  
